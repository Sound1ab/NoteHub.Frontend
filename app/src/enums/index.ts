export enum EXEC_COMMAND {
  UNDERLINE = 'underline',
}

export enum COLOR {
  DARK = 'dark',
  LIGHT = 'light',
  ACCENT = 'accent',
  MEDIUM = 'medium',
}

export enum LOCAL_STORAGE {
  KEY = 'noted::access_token',
}

export enum GRAPHQL {
  DEV_GRAPHQL = process.env.REACT_APP_GRAPHQL_DEV! as any,
  PROD_GRAPHQL = process.env.NODE_ENV! as any,
}

export enum APOLLO_ERRORS {
  UNAUTHENTICATED = 'UNAUTHENTICATED',
}
