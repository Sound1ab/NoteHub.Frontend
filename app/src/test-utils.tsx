import * as React from 'react'
import { render } from 'react-testing-library'
import { ThemeProvider } from './components/utility'

const customRender = (node: any, ...options: any[]) =>
  render(<ThemeProvider>{node}</ThemeProvider>, ...options)

// re-export everything
export * from 'react-testing-library'

// override render method
export { customRender as render }
