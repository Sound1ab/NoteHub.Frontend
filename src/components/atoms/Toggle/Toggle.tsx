import * as React from 'react'
import { ReactNode } from 'react'

import { ISwitch, Switch } from './Switch'

// this switch implements a checkbox input and is not relevant for this example
interface IContextProps {
  on: boolean
  toggle: () => void
}

const ToggleContext = React.createContext<Partial<IContextProps>>({})

function useEffectAfterMount(cb: () => void, otherDependencies: unknown[]) {
  const justMounted = React.useRef(true)

  const dependencies = [cb, ...otherDependencies]

  React.useEffect(() => {
    if (!justMounted.current) {
      return cb()
    }

    justMounted.current = false
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, dependencies)
}

interface IToggle {
  onToggle: (on: boolean) => void
  children: ReactNode
  initialState?: boolean
}

export function Toggle({ children, onToggle, initialState = false }: IToggle) {
  const [on, setOn] = React.useState(initialState)

  const toggle = React.useCallback(() => setOn((oldOn) => !oldOn), [])

  useEffectAfterMount(() => {
    onToggle(on)
  }, [on])

  const value = React.useMemo(() => ({ on, toggle }), [toggle, on])

  return (
    <ToggleContext.Provider value={value}>{children}</ToggleContext.Provider>
  )
}

function useToggleContext() {
  const context = React.useContext(ToggleContext)

  if (!context) {
    throw new Error(
      `Toggle compound components cannot be rendered outside the Toggle component`
    )
  }

  return context
}

export function ToggleButton({ id }: Pick<ISwitch, 'id'>) {
  const { on, toggle } = useToggleContext()

  return <Switch on={on} onClick={toggle} id={id} />
}
