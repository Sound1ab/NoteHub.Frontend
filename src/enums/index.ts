export enum TOGGLES {
  ROCK = 'rock',
  FOLK = 'folk',
  JAZZ = 'jazz',
  ALL = 'all',
}
