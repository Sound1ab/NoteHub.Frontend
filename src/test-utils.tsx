import '@testing-library/jest-dom/extend-expect'

import { EventType } from '@testing-library/dom/types/events'
import { act, fireEvent, render } from '@testing-library/react'
import React, { ReactNode } from 'react'

import {
  IconProvider,
  MockProvider,
  ThemeProvider,
} from './components/providers'
import { FileTreeProvider } from './components/templates/Dashboard/Sidebar/FileTree/FileTreeProvider'

export type FireObject = {
  [K in EventType]: (
    element: Document | Element | Window,
    options?: Record<string, unknown>
  ) => Promise<boolean>
}
global.matchMedia = () =>
  ({
    matches: false,
  } as MediaQueryList)

const wait = (): Promise<void> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve()
    }, 1)
  })
}

const Context = ({ node }: { node: ReactNode }) => (
  <MockProvider>
    <ThemeProvider>
      {() => (
        <IconProvider>
          <FileTreeProvider>{node}</FileTreeProvider>
        </IconProvider>
      )}
    </ThemeProvider>
  </MockProvider>
)

const customRender = async (
  node: ReactNode,
  { waitForLoad = true, ...options } = {}
) => {
  const { rerender, ...rest } = render(<Context node={node} />, options)

  if (waitForLoad) {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    await act(wait)
  }

  return {
    ...rest,
    rerender: async (node: ReactNode) => {
      rerender(<Context node={node} />)

      // eslint-disable-next-line @typescript-eslint/no-empty-function
      await act(wait)
    },
  }
}

const customFireEvent = Object.entries(fireEvent).reduce(
  (
    acc: Record<
      string,
      (
        element: Document | Element | Window,
        options: Record<string, string>
      ) => Promise<unknown>
    >,
    keyValue
  ) => {
    const [key, value] = keyValue
    acc[key] = async (
      element: Document | Element | Window,
      options: Record<string, unknown>
    ) => {
      if (typeof value === 'function') {
        value(element, options)
      }
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      await act(wait)
    }
    return acc
  },
  {}
)

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }

export { customFireEvent as fireEvent }
