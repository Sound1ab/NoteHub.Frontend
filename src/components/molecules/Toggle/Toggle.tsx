import React from 'react'
import { TOGGLES } from '../../../enums'
import { styled } from '../../../theme'
import { Heading } from '../../atoms'

const Style = styled.div`
  position: relative;
  width: 100%;
  border: 1px solid ${({ theme }) => theme.colors.brand};
  border-radius: 5px;
  display: flex;
  justify-content: space-between;

  h6 {
    color: ${({ theme }) => theme.colors.brand};
  }

  button + button {
    border-left: 1px solid ${({ theme }) => theme.colors.brand};
  }

  > button {
    flex: 1;
    display: flex;
    justify-content: center;
    padding: ${({ theme }) => theme.spacing.xs};
  }

  .toggle-active {
    background-color: ${({ theme }) => theme.colors.brand};
    h6 {
      color: ${({ theme }) => theme.colors.text.tertiary};
    }
  }
`

interface IToggle {
  toggles: TOGGLES[]
  activeToggle: TOGGLES
  setToggle: (toggle: TOGGLES, e: React.MouseEvent<HTMLElement>) => void
}

export function Toggle({ toggles, activeToggle, setToggle }: IToggle) {
  return (
    <Style>
      {toggles.map(toggle => (
        <button
          data-testid={toggle}
          key={toggle}
          className={activeToggle === toggle ? 'toggle-active' : ''}
          onClick={setToggle.bind(null, toggle)}
        >
          <Heading type="h6">{toggle}</Heading>
        </button>
      ))}
    </Style>
  )
}
