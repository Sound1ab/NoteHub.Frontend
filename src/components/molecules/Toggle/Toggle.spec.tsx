import React from 'react'
import { cleanup } from 'react-testing-library'
import { TOGGLES } from '../../../enums'
import { render } from '../../../test-utils'
import { Toggle } from './Toggle'

const MOCK_TOGGLE = TOGGLES.FOLK
const MOCK_SET_TOGGLE = jest.fn()

afterEach(cleanup)

describe('Toggle.tsx', () => {
  test('toggle matches snapshot', () => {
    const { asFragment } = render(
      <Toggle
        setToggle={MOCK_SET_TOGGLE}
        toggles={[TOGGLES.ROCK, TOGGLES.JAZZ, TOGGLES.FOLK, TOGGLES.ALL]}
        activeToggle={MOCK_TOGGLE}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('toggle highlights correct cell', () => {
    const { getByTestId } = render(
      <Toggle
        setToggle={MOCK_SET_TOGGLE}
        toggles={[TOGGLES.ROCK, TOGGLES.JAZZ, TOGGLES.FOLK, TOGGLES.ALL]}
        activeToggle={MOCK_TOGGLE}
      />
    )

    const button = getByTestId(MOCK_TOGGLE)
    expect(button.classList.contains('toggle-active')).toBe(true)
  })
})
