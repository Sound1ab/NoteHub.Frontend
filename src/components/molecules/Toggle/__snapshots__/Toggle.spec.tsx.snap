// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle.tsx toggle matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bxivhb hbJYnJ"
  >
    <button
      class=""
      data-testid="rock"
    >
      <h6
        class="sc-ifAKCX jpeJOs"
        type="h6"
      >
        rock
      </h6>
    </button>
    <button
      class=""
      data-testid="jazz"
    >
      <h6
        class="sc-EHOje fBVHip"
        type="h6"
      >
        jazz
      </h6>
    </button>
    <button
      class="toggle-active"
      data-testid="folk"
    >
      <h6
        class="sc-bZQynM jheleg"
        type="h6"
      >
        folk
      </h6>
    </button>
    <button
      class=""
      data-testid="all"
    >
      <h6
        class="sc-gzVnrw fDlMua"
        type="h6"
      >
        all
      </h6>
    </button>
  </div>
</DocumentFragment>
`;
